rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala

```diff
@@ -0,0 +1,317 @@
+    ): F[List[Trie]] =
+      partialTrie match {
+        case TriePath(Vector(), None, Nil) => // kickstart a trie
+          Applicative[F].pure(Skip(ByteVector(remainingPath), newLeaf) :: Nil)
+
+        case TriePath(Vector(Skip(skippedPath, LeafPointer(_))), None, common)
+            if common == skippedPath.toSeq.toList => // update 1 element trie
+          Applicative[F].pure(Skip(ByteVector(common), newLeaf) :: Nil)
+
+        case TriePath(Vector(), Some(Skip(affix, existingPointer)), Nil) => // split skip at root
+          val affixBytes                   = affix.toSeq.toList
+          val prefixPath                   = commonPrefix(affixBytes, remainingPath)
+          val incomingHead :: incomingTail = remainingPath.drop(prefixPath.size)
+          val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+          val (dividedElems, _) = divideSkip(
+            incomingTail,
+            incomingHead,
+            newLeaf,
+            existingTail,
+            existingHead,
+            existingPointer,
+            prefixPath
+          )
+          Applicative[F].pure(dividedElems)
+
+        case TriePath(
+            elems :+ (pointerBlock: PointerBlock),
+            Some(Skip(affix, existingPointer)),
+            common
+            ) => // split existing skip
+          val incomingPath                 = remainingPath.drop(common.size)
+          val affixBytes                   = affix.toSeq.toList
+          val prefixPath                   = commonPrefix(incomingPath, affixBytes)
+          val incomingHead :: incomingTail = incomingPath.drop(prefixPath.size)
+          val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+          val (dividedElems, dividedTopPtr) = divideSkip(
+            incomingTail,
+            incomingHead,
+            newLeaf,
+            existingTail,
+            existingHead,
+            existingPointer,
+            prefixPath
+          )
+          val updatedPointerBlock = pointerBlock.updated((toInt(common.last), dividedTopPtr) :: Nil)
+          for {
+            result <- rehash(common.init, elems, updatedPointerBlock, dividedElems)
+          } yield result
+
+        case TriePath(elems :+ (pastPb: PointerBlock), None, common) => // add to existing node
+          val key          = remainingPath.drop(common.size)
+          val (ptr, nodes) = skip(key, newLeaf)
+          val updatedPb    = pastPb.updated((toInt(common.last), ptr) :: Nil)
+          for {
+            result <- rehash(common.init, elems, updatedPb, nodes.toList)
+          } yield result
+
+        case TriePath(elems :+ Skip(affix, LeafPointer(_)), _, path) => // update value
+          rehash(path.dropRight(affix.size.toInt), elems, Skip(affix, newLeaf), Nil)
+        case _ => Sync[F].raiseError(MalformedTrieError)
+      }
+
+    private def rehash(
+        path: KeyPath,
+        rest: Vector[Trie],
+        lastSeen: Trie,
+        acc: List[Trie]
```

Comment:
acc is an accumulator only inside the go function.

This outside parameter should have a different name like initialSomething